src/attached_probe.cpp

```diff
@@ -811,48 +827,66 @@ void AttachedProbe::load_prog(BPFfeature &feature)
-
-        auto &btf = prog_.getBTF();
-        auto btf_fd = bpf_btf_load(btf.data(), btf.size(), &btf_opts);
-
-        opts.prog_btf_fd = btf_fd;
+        int btf_fd = -1;
+        if (load_config.btf) {
+          LIBBPF_OPTS(bpf_btf_load_opts,
+                      btf_opts,
+                      .log_buf = log_buf.get(),
+                      .log_level = static_cast<__u32>(log_level),
+                      .log_size = static_cast<__u32>(log_buf_size), );
+
+          auto &btf = prog_.getBTF();
+          btf_fd = bpf_btf_load(btf.data(), btf.size(), &btf_opts);
+          if (btf_fd < 0) {
+            logbuf << ": BTF load failed";
+            goto log_error; // and continue to the next load_config fallback
+          }
+          opts.prog_btf_fd = (__u32)btf_fd;
+        }
 
-        if (!func_infos.empty()) {
+        // Skip .BTF.ext func_info relocation on fallback for older kernels
+        if (load_config.func_infos && !func_infos.empty()) {
           opts.func_info_rec_size = sizeof(struct bpf_func_info);
           opts.func_info = func_infos.data();
           opts.func_info_cnt = func_infos.size() / sizeof(struct bpf_func_info);
         }
 
-        // Don't attempt to load the program if the BTF load failed.
-        // This will fall back to the error handling for failed program load,
-        // which is more robust.
-        if (btf_fd >= 0) {
+        // Load the BPF program on BTF load success or on fallback without it
+        if (!load_config.btf || btf_fd >= 0) {
           progfd_ = bpf_prog_load(static_cast<::bpf_prog_type>(prog_type),
                                   name.c_str(),
                                   license,
                                   reinterpret_cast<const struct bpf_insn *>(
                                       insns.data()),
                                   insns.size() / sizeof(struct bpf_insn),
-                                  &opts);
-          close(btf_fd);
+                                  &opts); // Configured opts are passed here
+          if (btf_fd >= 0)
+            close(btf_fd);
         }
       }
 
       if (opts.attach_btf_obj_fd > 0)
         close(opts.attach_btf_obj_fd);
       if (progfd_ >= 0)
         break;
-    }
+      logbuf << " = " << progfd_ << ": " << std::strerror(-progfd_);
+    log_error:
+      if (!logbuf.str().empty()) { // log the error of the failed attempt
+        LOG(V1) << logbuf.str();
+        logbuf.str("");
+      }
+    } // obsolete indendation level, has no effect, kept for a small commitdiff
+  } // for (auto &load_config : load_configs): handle kernel features & versions
+
+  if (!logbuf.str().empty()) {
+    LOG(V1) << logbuf.str() << ": Success";
   }
 
   if (progfd_ < 0) {
     if (bt_verbose) {
       std::cerr << std::endl
+                << "bpf_prog_load() failed with error " << progfd_ << ": "
+                << std::strerror(-progfd_) << std::endl
```

Comment:
Is this necessary now when the same is printed for each failed attempt?

Reply:
Thanks, correct, that is no longer necessary as it is already logged for each failed attempt!

Removed them! - I backed out these two lines from the commit, no other changes in this push.